File: 170.TwoSumIII-DataStructureDesign/TwoSumHashMap.java
class TwoSumHashMap {
  private HashMap<Integer, Integer> items;

  public TwoSum() {
    this.items = new HashMap<Integer, Integer>();
  }

  public void add(int number) {
    if (this.items.containsKey(number)) {
      this.items.replace(number, this.items.get(number) + 1);
    } else {
      this.items.put(number, 1);
    }
  }

  public boolean find(int value) {
    for (Map.Entry<Integer, Integer> entry : this.items.entrySet()) {
      int complement = value - entry.getKey();
      if (complement != entry.getKey()) {
        if (this.items.containsKey(complement))
          return true;
      } else {
        if (entry.getValue() > 1)
          return true;
      }
    }

    return false;
  }
}

File: 170.TwoSumIII-DataStructureDesign/two_sum_hashmap.cpp
#include <unordered_map>
using namespace std;

class TwoSum
{
private:
  unordered_map<int, int> items;

public:
  TwoSum() {}

  void add(int number)
  {
    if (items.find(number) != items.end())
    {
      items[number]++;
    }
    else
    {
      items[number] = 1;
    }
  }

  bool find(int value)
  {
    for (const auto &item : items)
    {
      int complement = value - item.first;

      if (complement != item.first)
      {
        if (items.find(complement) != items.end())
        {
          return true;
        }
      }
      else
      {
        if (item.second > 1)
        {
          return true;
        }
      }
    }
    return false;
  }
};